### PR TITLE
PS-6868 Merge fb-prod201906 (5.7)

### PR DESCRIPTION
**Pull request by inikep**
1. Update rocksdb submodule to [a78503bd6c80a3c4137df1962a972fe406b4d90b Temporarily disable snapshot list refresh for atomic flush stress test (#5581)](https://github.com/facebook/rocksdb/commit/a78503bd6c80a3c4137df1962a972fe406b4d90b) (v5.8-1722-ga78503bd6)

2. Add new MyRocks variables `rocksdb_table_stats_background_thread_nice_value`, `rocksdb_table_stats_max_num_rows_scanned`, `rocksdb_table_stats_recalc_threshold_count`,
 `rocksdb_table_stats_recalc_threshold_pct`, `rocksdb_table_stats_use_table_scan`, and `rocksdb_trace_block_cache_access`.

3. Cherry-pick the following commits from `fb-prod201906`:
- [801de516aeda87d757afedf84405b7eee35b7824 Remove caching of m_update_scope data across rows in single update](https://github.com/facebook/mysql-5.6/commit/801de516aeda87d757afedf84405b7eee35b7824)
- [4bdfb5557ae8bf47c03e053a341a9f06545e559c Improve statistics collection in MyRocks](https://github.com/facebook/mysql-5.6/commit/4bdfb5557ae8bf47c03e053a341a9f06545e559c)
- [7e67c47d6e227cecc2cfabd90a2f1f1890d8a630 Use std::unique_ptr instead of shared_ptr for Rdb_converter](https://github.com/facebook/mysql-5.6/commit/7e67c47d6e227cecc2cfabd90a2f1f1890d8a630)
- [c42081d12c66250360f295d68b3fca0a323d8b9c Support tracing block cache access. Reviewed](https://github.com/facebook/mysql-5.6/commit/c42081d12c66250360f295d68b3fca0a323d8b9c)

4. Null cherry-picked upstream commits:
- [df34666abb743015c6039e4a440cfbe10f645114 Fix rocksdb_hotbackup.gtid test](https://github.com/facebook/mysql-5.6/commit/df34666abb743015c6039e4a440cfbe10f645114)
- [a3076737ce3b403d30cedae955dc53cb7b470ef7 Ignore idempotent errors in TTL enabled tables](https://github.com/facebook/mysql-5.6/commit/a3076737ce3b403d30cedae955dc53cb7b470ef7)
- [eb2f1bdfb5233568d0a54d5e0a43185d75a7b9a4 Small fix in Rdb_tbl_def::has_ttl_col](https://github.com/facebook/mysql-5.6/commit/eb2f1bdfb5233568d0a54d5e0a43185d75a7b9a4)
- [2d9301955c92844dce6cef6d4f885de24fce7f1c Fix group-by plan with min/max aggregates and multiple infixes (#1039)](https://github.com/facebook/mysql-5.6/commit/2d9301955c92844dce6cef6d4f885de24fce7f1c)
- [3e767d3dc2c2144f3010732d1f2cd9de56443be8 MySQL bypass change #2 - SELECT parsing](https://github.com/facebook/mysql-5.6/commit/3e767d3dc2c2144f3010732d1f2cd9de56443be8)
- [d3f9187375aa32fa95179b60bf7c88dd67c483e3 Fix valgrind issue in bypass](https://github.com/facebook/mysql-5.6/commit/d3f9187375aa32fa95179b60bf7c88dd67c483e3)
- [2b0611ec0bed1d9223403efe56cb849c0388e864 Repurpose SLAVE_EVENT as METADATA_EVENT](https://github.com/facebook/mysql-5.6/commit/2b0611ec0bed1d9223403efe56cb849c0388e864)
- [0f4f1ab17f60c74c579c9066951bef829cd56b6d Revert "Fix writing binlogs with checksums during slave recovery"](https://github.com/facebook/mysql-5.6/commit/0f4f1ab17f60c74c579c9066951bef829cd56b6d)
- [88d0a0ee3a1aca408f5ddccc9d91d8b7cee0bf39 Assign HLC timestamp during ordered commit and write to binlog](https://github.com/facebook/mysql-5.6/commit/88d0a0ee3a1aca408f5ddccc9d91d8b7cee0bf39)
- [9f014d31c2066523459478a8e0345442a85d5fd2 Replacing boost ptree by rapidjson](https://github.com/facebook/mysql-5.6/commit/9f014d31c2066523459478a8e0345442a85d5fd2)
- [a0c025ce5c80fe3846bdc7510ca443bb7baffabc Update jemalloc stats only when needed](https://github.com/facebook/mysql-5.6/commit/a0c025ce5c80fe3846bdc7510ca443bb7baffabc)
- [9cc615d0f59d006bfe153e995522941e307d25a4 Persist max HLC from previous binlog into new binlog file](https://github.com/facebook/mysql-5.6/commit/9cc615d0f59d006bfe153e995522941e307d25a4)
- [a2e89c17fad3009bf13aba0c8e7dcc4310b47bb9 Initialize HLC during server restart](https://github.com/facebook/mysql-5.6/commit/a2e89c17fad3009bf13aba0c8e7dcc4310b47bb9)
- [7bfac75e24242b4bb6f06c796aaf7ba46d95f7c1 Return hlc timestamp for a write txn](https://github.com/facebook/mysql-5.6/commit/7bfac75e24242b4bb6f06c796aaf7ba46d95f7c1)
- [822bfe2cb935051b626db579d9e94731ae59f244 Stress tests for HLC feature](https://github.com/facebook/mysql-5.6/commit/822bfe2cb935051b626db579d9e94731ae59f244)
- [dde7f301d02d47bf25d8b6579927e703b9d8d30e add new type to metadata event](https://github.com/facebook/mysql-5.6/commit/dde7f301d02d47bf25d8b6579927e703b9d8d30e)
- [4fbe4c0e31d93b5265c8e64e744ebdf077e2941a Calling remove_slave() when plugin fails while starting dump thread](https://github.com/facebook/mysql-5.6/commit/4fbe4c0e31d93b5265c8e64e744ebdf077e2941a)
- [64b9e89d92513e45a525380fa6353a984ef22ea8 Fixing slave_check_before_image_consistency in MTS mode](https://github.com/facebook/mysql-5.6/commit/64b9e89d92513e45a525380fa6353a984ef22ea8)
- [85c6ed5ee398672aacf94a669f53c77bd9577b6d Fix partial trx condition when waiting for workers to finish](https://github.com/facebook/mysql-5.6/commit/85c6ed5ee398672aacf94a669f53c77bd9577b6d)
- [145292832b93cfc31a865474e0da1ac707d34b46 Better handle bad db_metadata value](https://github.com/facebook/mysql-5.6/commit/145292832b93cfc31a865474e0da1ac707d34b46)
- [22e4515d7c14924261dc61f7d58715ecf4410b77 reserve additional raft plugin related types in metadata event](https://github.com/facebook/mysql-5.6/commit/22e4515d7c14924261dc61f7d58715ecf4410b77)
- [0464fa528c2e3f78ec9ea4c5fa982f5e61bf1c25 Fixing an illegal mem ref bug caused by large trxs](https://github.com/facebook/mysql-5.6/commit/0464fa528c2e3f78ec9ea4c5fa982f5e61bf1c25)
- [a4a29878764bf1ea7d883234beefbf5b07f5371a return current hlc with snapshot](https://github.com/facebook/mysql-5.6/commit/a4a29878764bf1ea7d883234beefbf5b07f5371a)

### CODE REVIEW

**Comment by inikep** [comment]
Tested at https://ps57.cd.percona.com/job/percona-server-5.7-param/245

File: storage/rocksdb/ha_rocksdb.cc [hunk 1]
```diff
@@ -1706,13 +1729,55 @@ static MYSQL_SYSVAR_UINT(
     RDB_DEFAULT_TBL_STATS_SAMPLE_PCT, /* everything */ 0,
     /* max */ RDB_TBL_STATS_SAMPLE_PCT_MAX, 0);
 
+static MYSQL_SYSVAR_UINT(table_stats_recalc_threshold_pct,
+                         rocksdb_table_stats_recalc_threshold_pct,
```
Comment (george-lorch) [inline storage/rocksdb/ha_rocksdb.cc:1846]: Formatting here looks to be inconsistent, is it?

File: storage/rocksdb/ha_rocksdb.cc [hunk 2]
```diff
@@ -12443,16 +12885,203 @@ void Rdb_background_thread::run() {
...
+    RDB_MUTEX_LOCK_CHECK(m_signal_mutex);
+    if (m_killed) {
+      RDB_MUTEX_UNLOCK_CHECK(m_signal_mutex);
+      break;
+    }
+
+    timespec ts;
+    clock_gettime(CLOCK_REALTIME, &ts);
+
+    // Wait for 24 hours if the table scan based index calculation
```
Comment (george-lorch) [inline storage/rocksdb/ha_rocksdb.cc:13044]: Stalling a background/worker thread for 24 hours?  This doesn't look like a proper design.  We should look closer at this before taking in this commit.

**Review by george-lorch — CHANGES_REQUESTED** [review]

File: storage/rocksdb/ha_rocksdb.cc (hunk 2 above)
Comment (inikep) [reply storage/rocksdb/ha_rocksdb.cc:13044]: Everything is fine here. When `rocksdb_table_stats_use_table_scan` is switched to `ON` we have the following calls that will unblock `mysql_cond_timedwait(&m_signal_cond, &m_signal_mutex, &ts)`:
```
ha_rocksdb::write_row
ha_rocksdb::update_table_stats_if_needed
Rdb_index_stats_thread::add_index_stats_request
Rdb_index_stats_thread::signal
```
`Rdb_index_stats_thread::signal` is called also with `rocksdb_done_func()`.

Comment (inikep) [reply storage/rocksdb/ha_rocksdb.cc:13044]: Basically it causes the table scan based index calculation every 24 hours when `rocksdb_table_stats_use_table_scan` is `OFF`.

File: storage/rocksdb/ha_rocksdb.cc (hunk 1 above)
Comment (inikep) [reply storage/rocksdb/ha_rocksdb.cc:1846]: Maybe it is but my policy is to keep a minimal diff with upstream to minimize conflicts. Therefore I prefer to wait until upstream will fix/change it. 
It seems that they do `clang-formatting` for entire MyRocks every few months. In the meantime MyRocks could be not `clang-format` valid.

Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc:1846]: OK, this is not how we handled MyRocks in the past, we want _our_ tree to be as perfect as possible even if it causes some merge difficulties.  If we are sure that upstream reformats regularly, then we might consider relaxing.  Ideally upstream should catch format errors before accepting patches.

File: storage/rocksdb/ha_rocksdb.cc (hunk 2 above)
Comment (george-lorch) [reply storage/rocksdb/ha_rocksdb.cc:13044]: OK, makes sense.

**Review by george-lorch — APPROVED** [review]

**Pull request merged**